### PR TITLE
Fix memory leaks

### PR DESCRIPTION
**Pull request by hageboeck**
Fixed three leaks here:
- TSystem
- Metacling
- RooStats

Leaks before fixes (using `ASAN_OPTIONS=detect_leaks=1 LSAN_OPTIONS=max_leaks=-1 root.exe ...`):
[lsan_before.log](https://github.com/root-project/root/files/6326991/lsan_before.log)

After fixes:
[Don't have it yet, because after rebasing, #7903 and #6577 broke my asan build. There are leaks in cling, though]. 

@Axel-Naumann:
Please check the first two commits. One needed a shared_ptr or you need a proper copy constructor. I leave it to you to decide if the shared_ptr is the best choice here.

@guitargeek @lmoneta :
Since I waved the magic `unique_ptr` wand on ToyMCSampler anyway, I added a second commit to take it all the way.
I think it would be nice if you backported at least the fix of the leak (the smaller commit), and the larger one can just go in master for a better future. See #7890 regarding the ToyMCSampler leak and where backports are needed. I will leave it open.

Finding those leaks was a bit trickier, because RooFit uses a memory pool. To find those RooArgSet leaks,
- you switch off the mempool
- track down the leak
- switch it on again.

I added two commits mostly for @guitargeek with documentation for you to understand why things are a bit complicated here.

### CODE REVIEW

File: roofit/roostats/inc/RooStats/ToyMCSampler.h [hunk 1]
```diff
@@ -249,7 +249,7 @@ class ToyMCSampler: public TestStatSampler {
 
       // densities, snapshots, and test statistics to reweight to
       RooAbsPdf *fPdf; // model (can be alt or null)
-      const RooArgSet* fParametersForTestStat;
+      std::unique_ptr<const RooArgSet> fParametersForTestStat;
```
Comment (guitargeek) [inline roofit/roostats/inc/RooStats/ToyMCSampler.h:245]: Does ROOT know automatically how to do the schema evolution from raw pointer to `std::unique_ptr`?

**Review by guitargeek — COMMENTED** [review]
Thanks a lot Stephan for fixing this and for the explanations!

I have just a question about schema evolution. Other that that the changes in this PR make sense to me!

File: roofit/roostats/inc/RooStats/ToyMCSampler.h (hunk 1 above)
Comment (pcanal) [reply roofit/roostats/inc/RooStats/ToyMCSampler.h:245]: Yes :).  You might still need to update the ClassDef  version number if any.

Comment (guitargeek) [reply roofit/roostats/inc/RooStats/ToyMCSampler.h:245]: Thanks, good to know!

**Comment by guitargeek** [comment]
Thanks for bringing my attention to this MemPool business. For https://github.com/root-project/root/pull/7907, I benchmarked an example fit with toys from CMS, and the MemPoolForRooSets related functions actually took a lot of time there. Maybe it's just my build environment, but it would be interesting to see how the performance is without the mem pool enabled.

Adding a unique ID to the RooArgSets as you suggest in the documentation easy, but I guess finding all the pointer comparisons that would need to be replaced is hard :(

**Comment by guitargeek** [comment]
I profiled the CMS toy fit example:
* [with the mempool](https://rembserj.web.cern.ch/rembserj/cgi-bin/igprof-navigator/combine_example_2_new)
* [without the mempool](https://rembserj.web.cern.ch/rembserj/cgi-bin/igprof-navigator/combine_example_2_new_nomempool)

Actually it runs about 20 % faster without the memory pool. Is that expected? Can it be a problem that I'm profiling a debug build?

**Comment by eguiraud** [comment]
> Can it be a problem that I'm profiling a debug build?

`RelWithDebInfo` is ok, a `Debug` build on the other hand is not necessarily indicative of the performance of the corresponding optimized code.

**Comment by guitargeek** [comment]
Okay, then I have to make with comparisons again without `Debug` at some point!

**Review by guitargeek — APPROVED** [review]
Thanks again for the quick fix! All changes in this PR make sense to me.

File: roofit/roostats/inc/RooStats/ToyMCSampler.h (hunk 1 above)
Comment (hageboeck) [reply roofit/roostats/inc/RooStats/ToyMCSampler.h:245]: I was hoping that Philippe would say that. :-)

**Comment by hageboeck** [comment]
> I profiled the CMS toy fit example:
> 
> * [with the mempool](https://rembserj.web.cern.ch/rembserj/cgi-bin/igprof-navigator/combine_example_2_new)
> * [without the mempool](https://rembserj.web.cern.ch/rembserj/cgi-bin/igprof-navigator/combine_example_2_new_nomempool)
> 
> Actually it runs about 20 % faster without the memory pool. Is that expected? Can it be a problem that I'm profiling a debug build?

Yes, totally @guitargeek. The mempool alone already checks that you don't do double deletes when in debug mode.
There's *a lot* of other checks (e.g. checking the cached result against one that's force-evaluated on the fly) when you are in debug mode.

**Comment by lmoneta** [comment]
The PR looks good to me, and it is nice to move to use unique_ptr.

I have tested the PR with the code used  to reproduce  #7890 and unfortunately the memory leak is still there

File: core/metacling/src/TClingDataMemberInfo.h [hunk 2]
```diff
@@ -74,7 +75,7 @@ class TClingDataMemberInfo final : public TClingDeclInfo {
 private:
 
    cling::Interpreter    *fInterp;    // Cling interpreter, we do *not* own.
-   TClingClassInfo       *fClassInfo = nullptr; // ClassInfo for the decl context, for X<Float16_t> vs X<float>.
+   std::shared_ptr<TClingClassInfo> fClassInfo = nullptr; // ClassInfo for the decl context, for X<Float16_t> vs X<float>.
```
Comment (Axel-Naumann) [inline core/metacling/src/TClingDataMemberInfo.h:78]: I'd prefer https://github.com/root-project/root/pull/7915

**Review by Axel-Naumann — APPROVED** [review]
The `core` change looks great, thanks for spotting! For `metacling` I'd prefer https://github.com/root-project/root/pull/7915 - please drop that commit here. Other than that: approved :-)

**Pull request merged**